image-cropper/src/app/cropper/canvas/canvas.rs

```diff
@@ -0,0 +1,336 @@
+                    self.scale = scale.clone();
+                }
+                _ => (),
+            }
+        }
+        if let Some(event) = event.downcast_ref::<CanvasEvent>() {
+            match &event {
+                CanvasEvent::LeftMouseDownInCanvas {
+                    position,
+                    tool_type,
+                } => match tool_type {
+                    ToolType::Hand => self.hand_tool_pushed_position = Some(position.clone()),
+                    _ => (),
+                },
+                _ => (),
+            }
+        }
+        if let Some(event) = event.downcast_ref::<NamuiEvent>() {
+            match &event {
+                NamuiEvent::MouseUp(_) => {
+                    self.handle_hand_tool_up();
+                }
+                NamuiEvent::KeyDown(event) => match event.code {
+                    namui::Code::Digit1 | namui::Code::KeyH => {
+                        self.change_tool(ToolType::Hand);
+                    }
+
+                    namui::Code::Digit2 | namui::Code::KeyM => {
+                        self.change_tool(ToolType::RectSelection);
+                    }
+
+                    namui::Code::Digit3 | namui::Code::KeyL => {
+                        self.change_tool(ToolType::PolySelection);
+                    }
+                    _ => (),
+                },
+                _ => (),
+            }
+        }
+        self.tool.update(event);
+    }
+
+    pub fn render(&self, props: CanvasProps) -> RenderingTree {
+        let image_size = self.image.size();
+        let offset = self.offset.clone();
+        let scale = self.scale.clone();
+        let current_tool_type = self.tool.get_current_tool_type().clone();
+        let hand_tool_pushed_position = self.hand_tool_pushed_position.clone();
+
+        let scaled_image_size = Wh {
+            width: image_size.width * self.scale,
+            height: image_size.height * self.scale,
+        };
+        let scaled_offset = Xy {
+            x: offset.x * self.scale,
+            y: offset.y * self.scale,
+        };
+        render([
+            render_background(&props.wh).attach_event(|builder| {
+                builder
+                    .on_wheel(move |event| {
+                        let offset = offset.clone();
+                        let scale = scale.clone();
+                        let canvas_wh = props.wh.clone();
+                        let image_size = image_size.clone();
+
+                        let managers = namui::managers();
+                        let mouse_manager = &managers.mouse_manager;
+                        let mouse_position = mouse_manager.mouse_position();
+                        let canvas_xy = event
+                            .namui_context
+                            .get_rendering_tree_xy(event.target)
+                            .expect("failed to get canvas xy");
+                        let local_mouse_position = Xy {
```

Comment:
I think you can do `mouse_position - canvas_xy`

Reply:
`mouse_position` is Xy<**i16**>, but `canvas_xy` is Xy<**f32**>. So we need to change i16 to f32.
There seems to be no other handy, sexy way.
![image](https://user-images.githubusercontent.com/38313680/175246053-10ec8679-51ab-4382-b9d8-06f3c30d9380.png)

Reply:
got-it